### ISSUE TITLE
Recommend `if (nzchar(x))` instead of `if (nchar(x))`

### ISSUE THREAD

**Issue by Bisaloo** [jimhester/lintr#3058]
I am seeing some cases where automatic integer to bool conversion is used and `nchar(x)` is used directly in an `if` clause, where `nzchar()` would be both faster and clearer.

This should likely be part of `nzchar()` linter.

**Comment by MichaelChirico** [comment]
another way about this would be to block implicit coercion in if() statements, which is technically part of the guide:

https://style.tidyverse.org/syntax.html#if-statements

**Comment by Bisaloo** [comment]
How would you go about this? Having a list of known functions that don't return logicals? I think most common offenders are `length()`, `sum()`, `nrow()`, `nchar()`, etc.

**Comment by MichaelChirico** [comment]
yea, that's the main reason the linter doesn't exist yet -- hesitation about coming up with such a list 😅

that's a good start and probably covers at least 50%. I'll throw LLM at finding other candidates and quantifying their prevalence.

**Comment by MichaelChirico** [comment]
OK, here's the report summary. I would exclude `getOption()` and `attr()`.

## Results

We analyzed 50 codebases (14 R base packages, 35 CRAN/git packages, and internal packages/scripts).

We found **3,171 cases** of definite implicit coercion.

### Breakdown by Culprit (Top Functions/Operators)

| Culprit | Count | Percentage | Description / Risk |
| :--- | :--- | :--- | :--- |
| `length` | 2,541 | 80.13% | Common shorthand for `length(x) > 0`. Safe if not `NULL`, but bad style. |
| `NROW` | 203 | 6.40% | Similar to `length`, returns integer. Coerces `0` to `FALSE`. |
| `nrow` | 139 | 4.38% | Returns integer or `NULL`. `if (NULL)` triggers a runtime error. |
| `anyDuplicated` | 75 | 2.37% | Returns index of first duplicate (integer) or `0`. Should be `> 0`. |
| `bitwAnd` | 63 | 1.99% | Bitwise AND, returns integer. Should be `!= 0`. |
| `getOption` | 39 | 1.23% | Returns option value. If option is not set, returns `NULL`, causing runtime error. |
| `%%` (modulo) | 24 | 0.76% | Returns numeric. E.g., `if (x %% 2)` (0 is FALSE, others TRUE). |
| `sum` | 23 | 0.73% | Returns numeric. E.g., `if (sum(x))` (0 is FALSE). |
| `match` | 18 | 0.57% | Returns integer indices or `NA`. |
| `attr` | 17 | 0.54% | Returns attribute value. Risk of `NULL` if attribute missing. |
| `nchar` | 17 | 0.54% | Returns integer. |
| `ncol` | 3 | 0.09% | Returns integer or `NULL`. |
| `diff` | 2 | 0.06% | Returns numeric vector. |
| `grep` | 2 | 0.06% | Returns integer vector. Crashes with `integer(0)` if not found. |
| Others | 5 | 0.16% | `prod`, `tail`, `as.integer`, `%/%`, `-` |
| **Total** | **3,171** | **100.00%** | |

**Comment by Bisaloo** [comment]
Very subjective but I'm not a fan of `anyDuplicated()` return value and I can see the appeal of making it work like `anyNA()`. I wouldn't consider it terribly bad style to use implicit coercion here.

Some like `grep()` might still benefit from special treatment so we can give more targeted replacement recommendation.

**Comment by MichaelChirico** [comment]
Yea it's always sat oddly with me that `anyDuplicated()` and `anyNA()` return types differ. We'll want an option for users with custom functions like `MyLength()` or `obj@length()`, we could exclude `anyDuplicated()` from the "always-on" set like:

```r
implicit_coercion_linter <- function(coercers = "anyDuplicated") {
  all_coercers <- union(c("length", ..., "grep"), coercers)
  ...
}
```

(my own bias is to leave `anyDuplicated()` linting by default)
